python/lsst/ts/wep/task/calcZernikesNeuralTask.py

```
@@ -0,0 +1,1305 @@
+
+        Notes
+        -----
+        This method processes one side of the focal pair (intra OR extra).
+        In production, a single "exposure" typically contains both intra
+        and extra-focal donuts in different corners, but this method
+        expects an exposure with donut stamps from only one focal position.
+
+        The complete wavefront estimation workflow involves:
+        1. Processing intra-focal donuts separately (this method)
+        2. Processing extra-focal donuts separately (this method)
+        3. Combining results from both sides using AggregatorNet
+
+        The exposure should a single detector image that the TARTS models
+        are trained to process, not full frame images.
+
+        The difference between 'aggregatedZernikes' and 'rawZernikes':
+        - 'aggregatedZernikes' is the final aggregated result from TARTS's
+          AggregatorNet
+        - 'rawZernikes' contains individual per-donut predictions before
+          aggregation, useful for analyzing individual donut contributions
+        """
+        self.log.debug("Calculating Zernikes from exposure; defocalType='%s'", defocalType)
+        with torch.no_grad():
+            aggregatedZernikes = self.tarts.deploy_run(exposure)
+        # Convert PyTorch tensor to numpy array
+        aggregatedZernikes = self._toNumpy(aggregatedZernikes)
+        self.log.debug("TARTS Zernike coefficients prediction array shape: %s", np.shape(aggregatedZernikes))
+
+        # Ensure croppedImage is a numpy array before passing to function
+        croppedImageNp = self._toNumpy(self.tarts.cropped_image)
+
+        # Get TARTS internal data for more reliable field positions and SNR
+        tartsInternalData = self.tarts.get_internal_data()
+        self.log.debug("Retrieved %d donuts from TARTS internal data", len(tartsInternalData))
+        # Extract OOD scores if available; fallback to NaN when missing
+        # or when the OOD model is not loaded
+        try:
```

The except will only be raised if tartsInternalData is not a list of dicts. You may want to throw an error instead of having a np.nan default